components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/GitpodPortsService.kt

```diff
@@ -0,0 +1,81 @@
+// Copyright (c) 2022 Gitpod GmbH. All rights reserved.
+// Licensed under the GNU Affero General Public License (AGPL).
+// See License-AGPL.txt in the project root for license information.
+
+package io.gitpod.jetbrains.remote
+
+import com.intellij.openapi.diagnostic.thisLogger
+import com.jetbrains.rd.util.URI
+import org.apache.http.client.utils.URIBuilder
+import java.util.Optional
+import java.util.regex.Pattern
+
+class GitpodPortsService {
+    companion object {
+        /** Host used by forwarded ports on JetBrains Client. */
+        const val FORWARDED_PORT_HOST = "127.0.0.1"
+    }
+    private val hostToClientForwardedPortMap: MutableMap<Int, Int> = mutableMapOf()
+
+    fun isForwarded(hostPort: Int): Boolean = hostToClientForwardedPortMap.containsKey(hostPort)
+
+    private fun getForwardedPort(hostPort: Int): Optional<Int> = Optional.ofNullable(hostToClientForwardedPortMap[hostPort])
+
+    fun setForwardedPort(hostPort: Int, clientPort: Int) {
+        hostToClientForwardedPortMap[hostPort] = clientPort
+    }
+
+    fun removeForwardedPort(hostPort: Int) {
+        hostToClientForwardedPortMap.remove(hostPort)
+    }
+
+    fun getLocalHostUriFromHostPort(hostPort: Int): URI {
+        val optionalForwardedPort = getForwardedPort(hostPort)
+
+        val port = if (optionalForwardedPort.isPresent) {
+            optionalForwardedPort.get()
+        } else {
+            thisLogger().warn(
+                    "gitpod: Tried to get the forwarded port of $hostPort, which was not forwarded. " +
+                    "Returning $hostPort itself."
+            )
+            hostPort
+        }
+
+        return URIBuilder()
+                .setScheme("http")
+                .setHost(FORWARDED_PORT_HOST)
+                .setPort(port)
+                .build()
+    }
+
+    interface LocalHostUriMetadata {
+        val address: String
+        val port: Int
+    }
+
+    fun extractLocalHostUriMetaDataForPortMapping(uri: URI): Optional<LocalHostUriMetadata> {
+        if (uri.scheme != "http" && uri.scheme != "https") return Optional.empty()
+
+        val localhostMatch = Pattern.compile("^(localhost|127(?:\\.[0-9]+){0,2}\\.[0-9]+|0+(?:\\.0+){0,2}\\.0+|\\[(?:0*:)*?:?0*1?])(?::(\\d+))?\$").matcher(uri.authority)
+
+        if (!localhostMatch.find()) return Optional.empty()
+
+        var address = localhostMatch.group(1)
+        if (address.startsWith('[') && address.endsWith(']')) {
```

Comment:
How could this happen😂

Reply:
note: ipv6

Reply:
https://pl.kotl.in/ERWpCkfaW this case `"http://[::1]:8080"` is not correct? addr should be `::1` and port `8080`

Reply:
Nice catch! That proves this function deserves a unit test. Let me try it.

ℹ️ This function is a translation of this JavaScript function:

https://github.com/gitpod-io/openvscode-server/blob/8047dd89d347c79ca6e686dd1968f496b6da27bf/src/vs/platform/tunnel/common/tunnel.ts#L134-L151

Reply:
@akosyakov, that's a good question from @mustard-mh: why is it removing the last 2 characters instead of removing just the last one (`]`) [here](https://github.com/gitpod-io/openvscode-server/blob/8047dd89d347c79ca6e686dd1968f496b6da27bf/src/vs/platform/tunnel/common/tunnel.ts#L144)?

And wouldn't it be better to return the full hostname, like URL returns?
<img width="233" alt="image" src="https://user-images.githubusercontent.com/418083/183121556-ecf121c6-3707-4fb0-967b-9416c613791b.png">

Currently the code is returning `{address: "::", port: 8080}` for `http://[::1]:8080/`.

Should we change it (in Kotlin code from this PR) to return "[::1]"?

Reply:
I've added the unit test 🎉 

They can be run with `./gradlew test` (when inside `ide/jetbrains/backend-plugin` folder) or via IDE UI:

<img width="631" alt="image" src="https://user-images.githubusercontent.com/418083/183126778-980261c8-0693-4c0c-b202-44556e7ea8e4.png">

Reply:
I don't remember no, but looking at your client it does not seem to be matter. You don't use address anyway.

What you should understand here though, that the same port can be served on ip4 and ip6 at the same time as far as I remember then on client machine depending on the host there should be different ports too. I think it is worth to  create a follow-up issue to consider it, and discuss with JB how to handle such cases. But it should NOT block this PR.